Use errors and fmt package to join and wrap errors

boekkooi-terramate opened the pull request:
Good day!

First of thank you for maintaining and providing the watermill package!

I opened this PR because I noticed that the `hashicorp/go-multierror` and `github.com/pkg/errors` where added as an indirect dependency when using watermill.
As these dependency have native alternatives I created this PR.

Since golang 1.20 the go `errors` package supports `errors.Join` which replaces the need for the `hashicorp/go-multierror` package which was providing the same functionality. 
To wrap errors `fmt.Errorf` with the `%w` format specifier (introduced in golang 1.13) is now used instead of `errors.Wrap` which was provided by the `github.com/pkg/errors` package.

Please let me know what you think and if any changes are needed.
Have a great day and thank you for reviewing this PR!

Cheers,
Warnar

ccoVeille commented on components/cqrs/event_processor.go:
```diff
@@ -1,11 +1,10 @@
 
 import (
+	"errors"
 	stdErrors "errors"
```
Same

ccoVeille commented on components/cqrs/event_processor_group.go:
```diff
@@ -1,12 +1,12 @@
 
 import (
+	"errors"
 	stdErrors "errors"
```
Same

ccoVeille commented on components/cqrs/event_processor_group_test.go:
```diff
@@ -7,9 +7,9 @@ import (
 	"testing"
 	"time"
 
+	"errors"
```
Move this one in first block

ccoVeille commented on components/requestreply/handler.go:
```diff
@@ -3,9 +3,9 @@ package requestreply
 import (
 	"context"
 
+	"errors"
 	"github.com/ThreeDotsLabs/watermill/components/cqrs"
```
```suggestion
 	"context"
	"errors"

 	"github.com/ThreeDotsLabs/watermill/components/cqrs"
```

ccoVeille commented on dev/update-examples-deps/main.go:
```diff
@@ -24,16 +23,16 @@ func main() {
-				if err := updateWatermill(dir, file); err != nil {
-					panic(fmt.Sprintf("failed to update %s: %s", file, err))
-				}
+				//
+				//fmt.Println("update of", file, "@", dir)
+				//
+				//if err := updateDeps(dir, file); err != nil {
+				//	panic(err)
+				//}
+				//
+				//if err := updateWatermill(dir, file); err != nil {
+				//	panic(fmt.Sprintf("failed to update %s: %s", file, err))
+				//}
```
The fact this is commented lead me to think that either:
- it's a debug that shouldn't be there
- it's a code that is no longer needed

I would leave it like this

boekkooi-terramate replied:
Sorry about this change. I forgot to revert in and that's how it sneaked in.
I rebased the PR so this change is removed.

ccoVeille commented on internal/publisher/retry_test.go:
```diff
@@ -105,7 +106,7 @@ func TestRetryPublisher_Publish_too_many_retries(t *testing.T) {
 
 	assert.Equal(t, 1, cnpErr.Len(), "attempted to publish one message, expecting one error")
-	assert.Equal(t, errCouldNotPublish, errors.Cause(cnpErr.Reasons()[msg.UUID]))
+	assert.Equal(t, errCouldNotPublish, cnpErr.Reasons()[msg.UUID])
```
```suggestion
	assert.ErrorIs(t, cnpErr.Reasons()[msg.UUID], errCouldNotPublish)
```

ccoVeille commented on message/router/middleware/deduplicator.go:
```diff
@@ -11,8 +11,8 @@ import (
 	"sync"
 	"time"
 
+	"errors"
```
Same

ccoVeille commented on message/router/middleware/throttle_test.go:
```diff
@@ -5,7 +5,7 @@ import (
 	"time"
 
-	"github.com/pkg/errors"
+	"errors"
```
Same

ccoVeille commented on message/router_test.go:
```diff
@@ -7,7 +7,7 @@ import (
 	"time"
 
-	"github.com/pkg/errors"
+	"errors"
```
Same

ccoVeille commented on tools/mill/cmd/googlecloud.go:
```diff
@@ -10,12 +10,12 @@ import (
 	"google.golang.org/api/iterator"
 	"gopkg.in/yaml.v2"
+
+	"github.com/ThreeDotsLabs/watermill/tools/mill/cmd/internal"
```
Please regroup with

         "github.com/ThreeDotsLabs/watermill" 
          "github.com/ThreeDotsLabs/watermill-googlecloud/pkg/googlecloud"

ccoVeille reviewed (commented):
A few suggestions.

But don't get me wrong about the number of comments I posted, I love your changes

boekkooi-terramate commented on _examples/real-world-examples/receiving-webhooks/main.go:
```diff
@@ -3,13 +3,13 @@ package main
+	"errors"
 	"flag"
+	"fmt"
 	"io/ioutil"
```
Good spot. 
I added the replacements in https://github.com/ThreeDotsLabs/watermill/pull/460/commits/3dc9d723edd8882f20efab876f587f698cea105c

boekkooi-terramate commented:
Hey @ccoVeille 

Thank you for the code review. 
I think I managed to sort all the imports correctly now and also removed ioutil usage.

Let me know if I missed something or if some other change is needed.
Cheers

ccoVeille commented on components/metrics/http_test.go:
```diff
@@ -17,7 +17,7 @@ func TestCreateRegistryAndServeHTTP_metrics_endpoint(t *testing.T) {
 	err := reg.Register(collectors.NewBuildInfoCollector())
 	if err != nil {
-		t.Fatal(errors.Wrap(err, "registration of prometheus build info collector failed"))
+		t.Fatal(fmt.Errorf("registration of prometheus build info collector failed: %w", err))
```
Here the previous code is somehow bad.

And you are changing it to something better, but unfortunately still bad.

See

ccoVeille commented on components/metrics/http_test.go:
```diff
@@ -26,7 +26,7 @@ func TestCreateRegistryAndServeHTTP_metrics_endpoint(t *testing.T) {
 	}
 
 	if err != nil {
-		t.Fatal(errors.Wrap(err, "call to metrics endpoint failed"))
+		t.Fatal(fmt.Errorf("call to metrics endpoint failed: %w", err))
 	}
```
Same either Fatalf, or require.NoError

boekkooi-terramate commented:
Hey @ccoVeille 

I added commit https://github.com/ThreeDotsLabs/watermill/pull/460/commits/bbe870c442e400f33a98fca92c3cdefec8a7ecd9 to remove some `t.Fatal` and replaced them with testify `require`. 
(I hope) I also fixed all import sort order issues now.

Please have another look and let me know if I missed something (again) :smile:

ccoVeille commented:
I can help with adding golangci-lint checkers to avoid such issues and others

ccoVeille commented:
> I can help with adding golangci-lint checkers to avoid such issues and others

I opened #463 about it

m110 commented on message/router/middleware/recoverer.go:
```diff
@@ -26,7 +25,7 @@ func Recoverer(h message.HandlerFunc) message.HandlerFunc {
 
 		defer func() {
 			if r := recover(); r != nil || panicked {
-				err = errors.WithStack(RecoveredPanicError{V: r, Stacktrace: string(debug.Stack())})
```
This might be a bit of a breaking change, actually. 🤔 Perhaps we could somehow keep the stack trace here?

boekkooi-terramate replied:
I'm not sure how this really is a breaking change or not.
A `debug.Stack` call is used to get the stacktrace into `RecoveredPanicError`. (via a call to [`runtime.Callers`](https://pkg.go.dev/runtime#Callers)).
This seems to me like the same function that [`errors.WithStack`](https://github.com/pkg/errors/blob/master/errors.go#L145) provides as it's using `runtime.Callers` and then formatting it.

Now, should someone use reflection or have an interface to represent and error from the errors package then yes this PR could break things. 
Personally, I feel this change is not breaking the contract of the library as it's still returning the `error` interface.

ccoVeille commented:
I'm unsure why the PR was closed.

The code was OK. Is it because of the comment here?

https://github.com/ThreeDotsLabs/watermill/pull/460#discussion_r1735696498

This was a small part of your changes.

Even if there were a discussion about it, most of the code was OK to merge, no?

m110 commented:
@ccoVeille I agree. I'm not sure about the comment yet (I'm reviewing lots of Watermill PRs now, so I need some time), but the rest of the good looks good to me.

m110 commented:
Oh, I think the OP's account got deleted.

ccoVeille commented:
Oh OK. Then you can consider to reopen the PR.

I can do it of course, but with some delay maybe.

I hope the commits will stay live for a few days

m110 commented:
Sadly, it seems we can't reopen it, as the fork is gone. 🫨

ccoVeille commented:
Are you able to merge? You could revert then.

Quite dirty, but why not

ccoVeille commented:
You can use this as a maintainer maybe

https://github.blog/news-insights/product-news/change-the-base-branch-of-a-pull-request/

this would prefer to merge in master, you will have then a branch and you will be able to open a PR for it
